src/meta/src/rpc/server.rs

```diff
@@ -171,316 +153,283 @@ pub async fn rpc_serve_with_store<S: MetaStore>(
+        let (follower_shutdown_tx, follower_shutdown_rx) = OneChannel::<()>();
+        let follower_handle: Option<JoinHandle<()>> = if !node_is_leader(&leader_rx) {
+            let address_info_clone = address_info.clone();
+            Some(tokio::spawn(async move {
+                let _ = tracing::span!(tracing::Level::INFO, "follower services").enter();
+                start_follower_srv(
+                    svc_shutdown_rx_clone,
+                    follower_shutdown_rx,
+                    address_info_clone,
+                )
+                .await;
+            }))
+        } else {
+            None
         };
 
-        let (svc_shutdown_tx, mut svc_shutdown_rx) = tokio::sync::oneshot::channel::<()>();
+        // wait until this node becomes a leader
+        while !node_is_leader(&leader_rx) {
+            leader_rx.changed().await.expect("Leader sender dropped");
+        }
 
-        let join_handle = tokio::spawn(async move {
-            tokio::select! {
-                _ = tokio::signal::ctrl_c() => {}
-                _ = &mut svc_shutdown_rx => {
-                    shutdown_election.await;
-                }
+        // shut down follower svc if node used to be follower
+        if let Some(handle) = follower_handle {
+            match follower_shutdown_tx.send(()) {
+                Ok(_) => tracing::info!("Shutting down follower services"),
+                Err(_) => tracing::error!("Follower service receiver dropped"),
             }
-        });
+            // Wait until follower service is down
+            handle.await.unwrap();
+        }
 
-        // FIXME: Avoid using join_handler and just pass around shutdown_sender
-        // https://github.com/risingwavelabs/risingwave/pull/6771
-        return Ok((join_handle, svc_shutdown_tx));
-    }
+        let elect_coord = ElectionCoordination {
+            election_handle,
+            election_shutdown,
+        };
 
-    tracing::info!("Node initially elected as leader");
+        let current_leader = services_leader_rx.borrow().0.clone();
+        start_leader_srv(
+            meta_store,
+            address_info,
+            max_heartbeat_interval,
+            opts,
+            current_leader,
+            elect_coord,
+            svc_shutdown_rx,
+        )
+        .await
+        .expect("Unable to start leader services");
+    });
 
-    let env = MetaSrvEnv::<S>::new(opts, meta_store.clone(), current_leader_info).await;
-    let fragment_manager = Arc::new(FragmentManager::new(env.clone()).await.unwrap());
-    let meta_metrics = Arc::new(MetaMetrics::new());
-    let registry = meta_metrics.registry();
-    monitor_process(registry).unwrap();
+    Ok((join_handle, svc_shutdown_tx))
+}
 
-    let cluster_manager = Arc::new(
-        ClusterManager::new(env.clone(), max_heartbeat_interval)
-            .await
-            .unwrap(),
-    );
-    let heartbeat_srv = HeartbeatServiceImpl::new(cluster_manager.clone());
+mod tests {
```

Comment:
```suggestion
#[cfg(test)]
mod tests {
```
Adding `#[cfg(test)]` is enough, then you don't have to add it for all functions and variables.